File: TelerikUIExamples/CustomAnimationPieChart.m
//
//  CustomAnimationPieChart.m
//  TelerikUIExamples
//
//  Copyright (c) 2013 Telerik. All rights reserved.
//

#import "CustomAnimationPieChart.h"
#import <TelerikUI/TelerikUI.h>

@interface CustomAnimationPieChart() <TKChartDelegate>
@end

@implementation CustomAnimationPieChart
{
    TKChart *_chart;
}

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        [self addOption:@"Animate" selector:@selector(animate)];
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    _chart = [[TKChart alloc] initWithFrame:[self exampleBounds]];
    _chart.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    _chart.allowAnimations = YES;
    _chart.delegate = self;
    [self.view addSubview:_chart];
    
    NSMutableArray *array = [[NSMutableArray alloc] init];
    [array addObject:[[TKChartDataPoint alloc] initWithValue:@10 name:@"Apple" ]];
    [array addObject:[[TKChartDataPoint alloc] initWithValue:@20 name:@"Google" ]];
    [array addObject:[[TKChartDataPoint alloc] initWithValue:@30 name:@"Microsoft" ]];
    [array addObject:[[TKChartDataPoint alloc] initWithValue:@5 name:@"IBM" ]];
    [array addObject:[[TKChartDataPoint alloc] initWithValue:@8 name:@"Oracle" ]];
    
    TKChartPieSeries *series = [[TKChartPieSeries alloc] init];
    series.selectionMode = TKChartSeriesSelectionModeDataPoint;
    [_chart addSeries:series withItems:array];
}

- (void)animate
{
    [_chart animate];
}

#pragma mark TKChartDelegate

- (CAAnimation *)chart:(TKChart *)chart animationForSeries:(TKChartSeries *)series withState:(TKChartSeriesRenderState *)state inRect:(CGRect)rect
{
    CFTimeInterval duration = 0;
    NSMutableArray *animations = [[NSMutableArray alloc] init];
    for (int i = 0; i<state.points.count; i++) {
        NSString *pointKeyPath = [state animationKeyPathForPointAtIndex:i];
        
        NSString *keyPath = [NSString stringWithFormat:@"%@.distanceFromCenter", pointKeyPath];
        CAKeyframeAnimation *a = [CAKeyframeAnimation animationWithKeyPath:keyPath];
        a.values = @[ @50, @50, @0 ];
        a.keyTimes = @[ @0, @(i/(i+1.)), @1 ];
        a.duration = 0.3 * (i+1.1);
        [animations addObject:a];
        
        keyPath = [NSString stringWithFormat:@"%@.opacity", pointKeyPath];
        a = [CAKeyframeAnimation animationWithKeyPath:keyPath];
        a.values = @[ @0, @0, @1 ];
        a.keyTimes = @[ @0, @(i/(i+1.)), @1 ];
        a.duration = 0.3 * (i+1.1);
        [animations addObject:a];
        
        duration = a.duration;
    }
    CAAnimationGroup *g = [[CAAnimationGroup alloc] init];
    g.duration = duration;
    g.animations = animations;
    return g;
}

@end

File: TelerikUIExamplesInSwift/TelerikUIExamplesInSwift/CustomAnimationPieChart.swift
//
//  CustomAnimationPieChart.swift
//  Swift Examples
//
//  Copyright (c) 2014 Telerik. All rights reserved.
//

import Foundation
import QuartzCore

class CustomAnimationPieChart: TKExamplesExampleViewController, TKChartDelegate {
    
    let chart = TKChart()
    
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: NSBundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)

        self.addOption("Animate", action: animate)
    }

    required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        chart.frame = self.view.bounds
        chart.autoresizingMask = UIViewAutoresizing(rawValue: UIViewAutoresizing.FlexibleWidth.rawValue | UIViewAutoresizing.FlexibleHeight.rawValue)
        chart.allowAnimations = true
        chart.delegate = self
        self.view.addSubview(chart)
        
        var array = [TKChartDataPoint]()
        array.append(TKChartDataPoint(name: "Apple", value: 10))
        array.append(TKChartDataPoint(name: "Google", value: 20))
        array.append(TKChartDataPoint(name: "Microsoft", value: 30))
        array.append(TKChartDataPoint(name: "IBM", value: 5))
        array.append(TKChartDataPoint(name: "Oracle", value: 8))
        
        let series = TKChartPieSeries(items: array)
        series.selectionMode = TKChartSeriesSelectionMode.DataPoint
        chart.addSeries(series)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func animate()
    {
        chart.animate()
    }
    
    //MARK: - TKChartDelegate
    
    func chart(chart: TKChart, animationForSeries series: TKChartSeries, withState state: TKChartSeriesRenderState, inRect rect: CGRect) -> CAAnimation? {
        var duration = 0.0
        var animations = [CAAnimation]()
        for i in 0..<state.points.count() {
            let pointKeyPath = state.animationKeyPathForPointAtIndex(i)!
            let keyPath = NSString(format: "%@.distanceFromCenter", pointKeyPath)
            var a = CAKeyframeAnimation(keyPath: keyPath as String)
            let interval = 0.3*(Double(i)+1.1)
            
            a.values = [50, 50, 0]
            a.keyTimes = [0.0, (Double(i)/(Double(i)+1.0)), 1.0]
            a.duration = interval
            animations.append(a)
            
            a = CAKeyframeAnimation(keyPath: "\(pointKeyPath).opacity")
            a.values = [0, 0, 1]
            a.keyTimes = [0.0, (Double(i)/(Double(i)+1.0)), 1.0]
            a.duration = interval
            animations.append(a)
            
            duration = interval
        }
        let g = CAAnimationGroup()
        g.duration = duration
        g.animations = animations
        return g
    }
}